pyomo/common/multithread.py

```diff
@@ -0,0 +1,61 @@
+from collections import defaultdict
+from threading import get_ident, main_thread
+
+
+class MultiThreadWrapper():
+    """A python object proxy that wraps different instances for each thread.
+
+    This is useful for handling thread-safe access to singleton objects without
+    having to refactor the code that depends on them.
+
+    Note that instances of the wrapped object are reused if two threads share the same
+    identifier, because identifiers could be reused and are unique only for any given moment.
+    See [get_ident()](https://docs.python.org/3/library/threading.html#threading.get_ident) for more information.
+    """
+
+    def __init__(self, base):
+        self.__mtdict = defaultdict(base)
+
+    def __getattr__(self, attr):
+        id = get_ident()
+        return getattr(self.__mtdict[id], attr)
+    
+    def __setattr__(self, attr, value):
+        if attr == '_MultiThreadWrapper__mtdict':
+            object.__setattr__(self, attr, value)
```

Comment:
Why do you allow overwriting the `__mtdict` (but prevent it for the `main_thread`)?

Reply:
Because, inside the __init__ I'm doing `self.__mtdict = ...` which calls the __setattr__

Reply:
Right - that was my point above: you could use `object.__setattr__` in `__init__` and then make the attribute not settable here (following what you do with `main_thread`).